### PR TITLE
add: Migrate vit gradcam code + pytorch export format for classificat…

### PR DESCRIPTION
**Pull request by AlessandroPolidori**
## Summary

Features:

- Gradcam for Visual Transformers in Classification and SklearnClassification tasks using Attention Gradient Rollout!
- Now Classification and SklearnClassification tasks can export models also in pytorch format
- This format is used by their respective evaluation/test tasks, so you can compute gradcams also when you test an already trained model. Other tasks currently export only in torchscript format and their respective eval. tasks can only import torchscript models.

## Type of Change

- New feature (non-breaking change that adds functionality)

## Checklist

Please confirm that the following tasks have been completed:

- [x] I have tested my changes locally and they work as expected. (Please describe the tests you performed.)
- [x] I have added unit tests for my changes, or updated existing tests if necessary.
- [x] I have updated the documentation, if applicable.
- [x] I have installed pre-commit and run locally for my code changes.

### CODE REVIEW

File: quadra/tasks/classification.py [hunk 1]
```diff
@@ -910,17 +1036,36 @@ def test(self) -> None:
...
-                predicted_classes.append(preds.tolist())
-                image_labels.extend(target.tolist())
+        grayscale_cams_list = []
+
+        if self.gradcam:
+            self.prepare_gradcam()
+
+        for batch_item in tqdm(test_dataloader):
+            im, target = batch_item
+            im = im.to(self.device).detach()
```
Comment (lorenzomammana) [inline quadra/tasks/classification.py:1046]: I don't like that we compute gradients when is not necessary

with torch.set_grad_enabled(self.gradcam):

File: quadra/utils/export.py [hunk 2]
```diff
@@ -59,6 +59,27 @@ def export_torchscript_model(
...
+        If the model is exported successfully, the path to the model is returned.
+
+    """
+    os.makedirs(output_path, exist_ok=True)
+    model.eval()
+    model.cpu()
+    model_path = os.path.join(output_path, model_name)
+    torch.save(model.state_dict(), model_path)
+    log.info("Pytorch model saved to %s", os.path.join(output_path, model_name))
+    return os.path.join(os.getcwd(), model_path)
```
Comment (lorenzomammana) [inline quadra/utils/export.py:80]: Empty line before return

File: quadra/utils/vit_explainability.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+
+def rollout(attentions: List[torch.Tensor], discard_ratio: float = 0.9, head_fusion: str = "mean") -> np.ndarray:
+    """Apply rollout on Attention matrices.
+
+    Args:
+        attentions: List of Attention matrices coming from different blocks
+        discard_ratio: Percentage of elements to discard
+        head_fusion: Strategy of fusion of attention heads
+    Returns:
```
Comment (lorenzomammana) [inline quadra/utils/vit_explainability.py:15]: Empty line before Returns

File: quadra/utils/vit_explainability.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+            a = a / a.sum(dim=-1).unsqueeze(1)
+            result = torch.matmul(a, result)
+    # Look at the total attention between the class token and the image patches
+    mask = result[:, 0, 1:]
+    batch_size = mask.size(0)
+    # In case of 224x224 image, this brings us from 196 to 14
+    width = int(mask.size(-1) ** 0.5)
+    mask = mask.reshape(batch_size, width, width).numpy()
+    mask = mask / mask.max(axis=(1, 2), keepdims=True)
+    return mask
```
Comment (lorenzomammana) [inline quadra/utils/vit_explainability.py:45]: White line before

File: quadra/utils/vit_explainability.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        device: The device to use. Defaults to "cpu".
+    """
+
+    def __init__(self, backbone: torch.nn.Module, lr_classifier: ClassifierMixin, device: torch.device):
+        super().__init__()
+        self.device = device
+        self.backbone = backbone.to(device)
+        self.num_classes = len(lr_classifier.classes_)
+        self.lr_classifier = lr_classifier
+        self.example_input_array = torch.rand(1, 3, 224, 224).to(device)  # TODO: Should not be hardcoded
```
Comment (lorenzomammana) [inline quadra/utils/vit_explainability.py:273]: 🚒

File: quadra/utils/vit_explainability.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+            targets_list = output.argmax(dim=1)
+        class_mask[torch.arange(output.shape[0]), targets_list] = 1
+        loss = (output * class_mask).sum()
+        loss.backward()
+        out = grad_rollout(self.attentions, self.attention_gradients, self.discard_ratio)
+
+        return out
+
+
+class ClassifierPytorchWrapper(torch.nn.Module):
```
Comment (lorenzomammana) [inline quadra/utils/vit_explainability.py:257]: Is this working from anything different than a logistic regression? 

If not I would check it before

**Review by lorenzomammana — CHANGES_REQUESTED** [review]
There are still some points I would review

File: quadra/modules/classification/base.py [hunk 7]
```diff
@@ -158,14 +191,22 @@ def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> A
                 grayscale_cam: gray scale gradcams
         """
         im, _ = batch
-        # inference_mode set to false because gradcam needs gradients
         if self.gradcam:
+            outputs = self(im)
```
[the remark this answers is not in the document]
Comment (AlessandroPolidori) [reply quadra/modules/classification/base.py:195]: i moved the common lines before the if statement btw

**Review by lorenzomammana — CHANGES_REQUESTED** [review]
Almost there 🔜

File: quadra/utils/vit_explainability.py (hunk 6 above)
Comment (AlessandroPolidori) [reply quadra/utils/vit_explainability.py:262]: Thinking about it, maybe it is suitable for sklearn.linear_model class of models only. I will verify today

**Review by lorenzomammana — APPROVED** [review]
LGTM
